casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -142,11 +142,25 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Capture: ConnectionsCell: Tr
                             )
       newFinalizedBlock <- maybeFinalizedChild match {
                             case Some(finalizedChild) =>
-                              updateLastFinalizedBlock(dag, finalizedChild)
+                              removeDeploysInFinalizedBlock(finalizedChild) *> updateLastFinalizedBlock(
+                                dag,
+                                finalizedChild
+                              )
                             case None => lastFinalizedBlockHash.pure[F]
                           }
     } yield newFinalizedBlock
 
+  private def removeDeploysInFinalizedBlock(finalizedChild: BlockHash): F[Unit] =
+    for {
+      b              <- ProtoUtil.unsafeGetBlock[F](finalizedChild)
+      deploys        = b.body.get.deploys.map(_.deploy.get)
+      deploysRemoved = deploys.count(deployHist.remove)
```

Comment:
```suggestion
      deploysRemoved = deploys.toList.foldM(0)((count, deploy) =>
        Monad[F].ifM(Sync[F].delay { deployHist.remove(deploy) })((count + 1).pure[F], count.pure[F])
      )
```
Or something similar as `deployHist.remove` is effectful